Sources/Services/ContainerAPIService/Containers/ContainersService.swift

```diff
@@ -163,111 +207,289 @@ public actor ContainersService {
+        do {
+            try await self.lock.withLock { context in
+                var item = try await self.get(id: id, context: context)
+                let runtime = item.snapshot.configuration.runtimeHandler
+                let sandboxClient = try await SandboxClient.create(
+                    id: id,
+                    runtime: runtime
+                )
+                try await sandboxClient.bootstrap(stdio: stdio)
+
+                try await self.exitMonitor.registerProcess(
+                    id: id,
+                    onExit: self.onContainerExit
+                )
+
+                item.client = sandboxClient
+                await self.setContainer(id, item, context: context)
+            }
+        } catch {
+            do {
+                try await _cleanup(id: id)
+            } catch {
+                self.log.error("failed to cleanup container \(id) after bootstrap failure: \(error)")
+            }
+            throw error
+        }
     }
 
-    private static func registerService(
-        plugin: Plugin,
-        loader: PluginLoader,
-        configuration: ContainerConfiguration,
-        path: URL
-    ) throws {
-        let args = [
-            "--root", path.path,
-            "--uuid", configuration.id,
-            "--debug",
-        ]
-        try loader.registerWithLaunchd(
-            plugin: plugin,
-            pluginStateRoot: path,
-            args: args,
-            instanceId: configuration.id
-        )
+    /// Create a new process in the container.
+    public func createProcess(
+        id: String,
+        processID: String,
+        config: ProcessConfiguration,
+        stdio: [FileHandle?]
+    ) async throws {
+        self.log.debug("\(#function)")
+
+        let item = try self._get(id: id)
+        do {
+            let client = try item.getClient()
+            try await client.createProcess(
+                processID,
+                config: config,
+                stdio: stdio
+            )
+        } catch {
+            do {
+                try await _cleanup(id: id)
+            } catch {
+                self.log.error("failed to cleanup container \(id) after start failure: \(error)")
+            }
+            throw error
+        }
     }
 
-    private func get(id: String, context: AsyncLock.Context) throws -> ContainerSnapshot {
-        try self._get(id: id)
+    /// Start a process in a container. This can either be a process created via
+    /// createProcess, or the init process of the container which requires
+    /// id == processID.
+    public func startProcess(id: String, processID: String) async throws {
+        self.log.debug("\(#function)")
+
+        do {
+            try await self.lock.withLock { context in
+                var item = try await self.get(id: id, context: context)
+                let client = try item.getClient()
+                try await client.startProcess(processID)
+
+                if id == processID {
```

Comment:
We only use it in two places it appears, but the code might read better if we had an `isInitProcess(containerId, processId)`